dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/PersistablesFilter.java

```diff
@@ -0,0 +1,412 @@
+import org.hisp.dhis.tracker.domain.Event;
+import org.hisp.dhis.tracker.domain.Relationship;
+import org.hisp.dhis.tracker.domain.RelationshipItem;
+import org.hisp.dhis.tracker.domain.TrackedEntity;
+import org.hisp.dhis.tracker.domain.TrackerDto;
+import org.hisp.dhis.tracker.preheat.TrackerPreheat;
+import org.hisp.dhis.tracker.report.TrackerErrorCode;
+import org.hisp.dhis.tracker.report.TrackerErrorReport;
+
+/**
+ * Determines whether entities can be persisted (created, updated, deleted)
+ * taking into account the {@link TrackerImportStrategy} and the parent/child
+ * relationships between entities. For example during
+ * {@link TrackerImportStrategy#CREATE} no valid child of an invalid parent can
+ * be created (i.e. enrollment of trackedEntity or event of enrollment). During
+ * {@link TrackerImportStrategy#UPDATE} a valid child of an invalid parent can
+ * be updated if the parent exists.
+ * <p>
+ * The {@link Result} returned from
+ * {@link #filter(TrackerBundle, EnumMap, TrackerImportStrategy)} can be trusted
+ * to only contain persistable entities. The {@link TrackerBundle} is not
+ * mutated!
+ * </p>
+ * <p>
+ * Errors are only added to {@link Result#errors} if they add information. For
+ * example a valid entity with invalid children cannot be deleted because of its
+ * children. Since the valid parent did not already have an error added during
+ * validation on will be added here. For more information see
+ * {@link #addErrorsForParents(List, TrackerDto)} and
+ * {@link #addErrorsForChildren(List, TrackerDto)}.
+ * </p>
+ * <p>
+ * This filter relies on preprocessing and all
+ * {@link org.hisp.dhis.tracker.validation.TrackerValidationHook}s having run
+ * beforehand. The following are some assumptions the filtering relies on:
+ * </p>
+ * <ul>
+ * <li>This does not validate whether the {@link TrackerImportStrategy} matches
+ * the state of a given entity. In case a non existing entity is updated it is
+ * expected to already be flagged in the {@code invalidEntities}.</li>
+ * <li>Existence is only checked in relation to a parent or child. During
+ * {@link TrackerImportStrategy#UPDATE} a valid enrollment can be updated if its
+ * parent the TEI is invalid but exists. Same applies to
+ * {@link TrackerImportStrategy#DELETE} as you cannot delete an entity that does
+ * not yet exist.</li>
+ * <li>An {@link Event} in an event program does not have an
+ * {@link Event#getEnrollment()} (parent) set in the payload. This expects one
+ * to be set during preprocessing. So events in program with or without
+ * registration do not get any special treatment.</li>
+ * <li>{@link TrackerImportStrategy#DELETE} with
+ * {@link org.hisp.dhis.security.Authorities#F_TEI_CASCADE_DELETE} is not
+ * treated differently than a delete without such authority. Validation should
+ * have already flagged entities in {@code invalidEntities} if they have
+ * children that cannot be deleted by that user.</li>
+ * </ul>
+ */
+class PersistablesFilter
```

Comment:
Maybe the delete and createOrUpdate logic all in one place can be hard to follow in the code.
I found myself to always having to remember what mark was meaning in the context I was following.

Do you think it would make sense to have them separated from the beginning, maybe even having 2 different filters one for the delete and the other for the createOrUpdate?